loki/transformations/pragma_model.py

```diff
@@ -0,0 +1,324 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import inspect
+
+from loki.batch import Transformation, ProcedureItem, ModuleItem
+from loki.ir import (
+    FindNodes, Pragma, Transformer, get_pragma_start_and_parameters
+)
+
+__all__ = ['PragmaModelTransformation']
+
+
+class GenericPragmaMapper:
+    """
+    A generic pragma mapper class.
+
+    Pragmas in the form
+
+    ``!$loki command-optionally-with-hyphen [param] [param_with_val(val)]``
+
+    get a visitor/handler method that looks like
+
+    .. code-block::
+       def visit_command_optionally_with_hyphen(self, pragma, [**kwargs]):
+           pass
+
+    The handler is responsible for returning either None or the updated
+    pragma.
+    """
+
+    def __init__(self):
+        handlers = {}
+        prefix = "pmap_"
+        for (name, meth) in inspect.getmembers(self, predicate=inspect.ismethod):
+            if not name.startswith(prefix):
+                continue
+            argspec = inspect.getfullargspec(meth)
+            if len(argspec.args) < 2:
+                raise RuntimeError("Visit method signature must be "
+                                   "pmap_foo(self, pragma, [**kwargs])")
+            handlers[name[len(prefix):]] = meth
+        self._handlers = handlers
+
+    def lookup_method(self, starts_with):
+        try:
+            return self._handlers[starts_with.lower()]
+        except KeyError:
+            return None
+
+    @classmethod
+    def default_retval(cls):
+        """
+        Default return value for handler methods.
+
+        Returns
+        -------
+        None
+        """
+        return None
+
+    def pmap(self, pragma, **kwargs): # pylint: disable=unused-argument
+        starts_with, parameters = get_pragma_start_and_parameters(pragma)
+        meth = self.lookup_method(starts_with.lower().replace('-', '_'))
+        print(f"map to meth: {meth} | starts_with: {starts_with}")
+        if meth is not None:
+            return meth(pragma, parameters, **kwargs)
+        return self.default_retval() # pragma # or return None
+
+
+class OpenACCPragmaMapper(GenericPragmaMapper):
+    """
+    Loki generic pragmas to OpenACC mapper.
+    """
+
+    def pmap_create(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        if param_device := parameters.get('device'):
+            return Pragma(keyword='acc', content=f'declare create({param_device})')
+        return self.default_retval()
```

Comment:
The default_retval case is never hit in tests, and I'm wondering if we shouldn't better raise an exception than returning something if we have a Loki directive that doesn't match the "spec".